src/Pact/Gas/Table.hs

```diff
@@ -0,0 +1,173 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE LambdaCase #-}
+module Pact.Gas.Table where
+
+import qualified Data.Map as Map
+
+import Pact.Types.Gas
+import Data.Map (Map)
+--import qualified Data.Map as M
+import Data.Text (Text)
+import qualified Data.Text as T
+
+import Pact.Types.Term
+import Pact.Types.Persistence
+
+-- NB: If pact ends up not having any variadic primitives (currently, I didn't spot any thus far, but also the types don't rule it out)
+-- then the primTableUnreduced here could lose the [Term Ref] function parameter, and simply be Map Text Gas.
+-- Attributing gas costs to Term Ref on its own is problematic in general -- typically we want to account for those costs when the terms
+-- become evaluated, but there may be some cases where the ability to inspect the argument list is helpful, so I'll leave this in for now.
+
+-- The argument lists in the GReduced case may be more useful, but I wasn't able to find instances where GReduced was in present use,
+-- so it's possible we want to simplify that away.
```

Comment:
Actually, I agreer with the bit about `[Term Ref]` not being the most suitable thing for calculating gas. Was there a motivation for doing it this way, or was this just intended to be an intermediary step for later?

Reply:
Ah, I left that comment in....

Well, in general the goal is to fail as early as possible, so I suppose that ideally you'd like to know about gas costs before you start evaluating the arguments to a primitive. If you do that, then you don't have the evaluated terms available when determining the cost. I removed GReduced because we literally never used it, but it would be possible to add it back in easily if we ever did want to evaluate the arguments first and then determine gas cost (which seems like it would at least be more precise).

I've thought a bunch about how we might be able to tidy up the whole situation with the different types for native operations (defRNative vs. defNative etc.) but haven't yet come to a solid conclusion on what would actually be right for that. I think something is needed, a lot of the case analysis required while implementing the primitives is pretty annoying.

Reply:
@cgibbard would love your thoughts on the `defXXXNative` mess; the pact builtins are a poorly-embedded DSL that require a lot of repeated patterns to be effective. 

@emilypi @cgibbard  Reg `[Term Ref]` the only function currently is to disambiguate overloaded invocations but even this is unreliable as they haven't been validated yet by the definition. I think we can go ahead and remove them as they are not inspectable in any meaningful way.